mlapi/discriminating_algo.py

```diff
@@ -0,0 +1,129 @@
+import statistics
+
+class DiscriminatingAlgo(object):
+    # Discrimination parameters
+    facets_sample_count = 3  # minimum docs by facet to be accepted
+    tolerance_interval = 10 # max standard deviation accepted ()
+    nbr_facet_values = 2 # to consider a facet as having a large number of values
+
+    facets = [] # names
+    step2 = {}
+    step3 = {}
+
+    def discriminating_facets_algo(self, facets_by_document):
+        documents_by_facet = self.inverse_dictionary(facets_by_document)
+        data = self.get_formated_data(documents_by_facet)
+        self.execute_algorithm(data)
+        if not (len(self.step3)):
+            self.adjust_parameters()
+            self.execute_algorithm(data)
+        return self.get_values_by_facet(self.step3)
+
+    def inverse_dictionary(self, facets_by_document):
+        documents_by_facet = {}
+        for k, v in facets_by_document.items():
+            for value in v:
+                documents_by_facet[value] = documents_by_facet.get(value, [])
+                documents_by_facet[value].append(k)
+        return documents_by_facet
+
+    def get_formated_data(self, documents_by_facet):
+        data = {}
+        list = []
+        for facet, docs in documents_by_facet.items():
+            t = (facet.name, facet.value)
+            if t not in list:
+                tuple = (facet.name, facet.value)
+                data[tuple] = docs
+                list.append(tuple)
+            elif t in list:
+                data[tuple].append(docs)
+        return data
+
+    def get_facet_names(self, data):
```

Comment:
Pythonic way to do the same thing is something like:

return [facet[0] for facet in data.keys() if facet[0] not in list]

https://stackoverflow.com/questions/6981717/pythonic-way-to-combine-for-loop-and-if-statement

Reply:
I used the if statement just to not to add duplicated facets.
So, I changed the list to a set to have only unique facet names. And then I convert it to a list
    def get_facet_names(self, data):
        return list({facet[0] for facet in data.keys()})